Week3/homework/step4-bonus.js

```diff
@@ -1,10 +1,61 @@
 'use strict';
 
-const values = ['a', 'b', 'c', 'd', 'a', 'e', 'f', 'c'];
+{
+  const values = ['a', 'b', 'c', 'd', 'a', 'e', 'f', 'c'];
 
-// Add your function here. Try and come up with a good name for this function
+  function removeDuplicates(array) {
+    const unique = [];
+    for (let i = 0; i < array.length; i++) {
+      if (unique.indexOf(array[i]) === -1) {
+        unique.push(array[i]);
+      }
+    }
+    return unique;
+  }
 
-// Replace `yourFunction` with the name of the function you just created
-const uniqueValues = yourFunction(values);
+  const uniqueValues = removeDuplicates(values);
 
-console.log(uniqueValues);
+  console.log(uniqueValues);
+}
+
+{
+  const values = ['a', 'b', 'c', 'd', 'a', 'e', 'f', 'c'];
+
+  function removeDuplicates(array) {
+    return array.filter(function (val) {
+      return !this[val] ? this[val] = true : false;
+    }, {});
+  }
+  const uniqueValues = removeDuplicates(values);
+
+  console.log(uniqueValues);
+}
+
+{
+  const values = ['a', 'b', 'c', 'd', 'a', 'e', 'f', 'c'];
+  function removeDuplicates() {
+    const unique = {};
+    values.forEach(function (i) {
+      if (!unique[i]) {
+        unique[i] = true;
+      }
+    });
+    return Object.keys(unique);
+  }
+
+  const uniqueValues = removeDuplicates(values);
+
+  console.log(uniqueValues);
+
+}
+
+{
+  const values = ['a', 'b', 'c', 'd', 'a', 'e', 'f', 'c'];
+
+  const unique = [...new Set(values)];
+
+  console.log(unique);
+}
+
+
+// Note: This step was made with EXTREME googling :)
```

Comment:
We developers google a lot to help us write our code. However, you must always be able to explain the code that you are using. Otherwise we're doing a sort of bismillah-type of programming. Can you explain all these versions yourself?

Which version is the most easy to explain for yourself?